### PR TITLE
HLP testing tool implementation

### PR DESCRIPTION
**Pull request by palaciosjeremias**
|Related issue|
|---|
|#1254|

The following PR adds a CLI tool to execute the HLP library with the intention of exploratory testing.
The tool accepts a list of 2 arguments as the LogQL expression and the event to be parsed with this expression.
Also accepts 1 argument as a yaml file with a list of expressions and events.
Usage:
```
 ./hlp_cli "FILENAME"
 ./hlp_cli "LOGQL_EXPRESSION" "EVENT"
```

Example input file:
```
-
  logql_expression: "<_map/MAP/ /=/.> <_dummy>"
  event: "key1=Value1 Key2=Value2. dummy"
-
  logql_expression: "<_map/MAP/ /=>"
  event: "key1=Value1 Key2="

-
  logql_expression: "<_json1/JSON> - <_json2/JSON>"
  event: "{\"String\":\"This is a string\"} - {\"String\":\"This is another string\"}"
-
  logql_expression: "<_> <_temp> <_temp1/type> <_temp2/type/type2>"
  event: "one temp temp1 temp2="
-
  logql_expression: "this is an url <url> in text"
  event: "this is an url https://user:password@wazuh.com:8080/path?query=%22a%20query%20with%20a%20space%22#fragment in text"
-
  logql_expression: "<source.ip> - <server.ip> -- <source.nat.ip> \"-\" \"-\""
  event: "127.0.0.1 - 2001:db8:3333:4444:CCCC:DDDD:EEEE:FFFF -- 0:db8:0:4:CCCC:0:EEEE:FFFF \"-\" \"-\""
```

Example output:
```
----------
LOGQL_EXPRESSION:
<_map/MAP/ /=/.> <_dummy>
EVENT:
key1=Value1 Key2=Value2. dummy
RESULT:
_dummy dummy
_map {"key1":"Value1","Key2":"Value2"}
----------

----------
LOGQL_EXPRESSION:
<_map/MAP/ /=>
EVENT:
key1=Value1 Key2=
RESULT:
----------

----------
LOGQL_EXPRESSION:
<_json1/JSON> - <_json2/JSON>
EVENT:
{"String":"This is a string"} - {"String":"This is another string"}
RESULT:
_json2 {"String":"This is another string"}
_json1 {"String":"This is a string"}
----------

----------
LOGQL_EXPRESSION:
<_> <_temp> <_temp1/type> <_temp2/type/type2>
EVENT:
one temp temp1 temp2=
RESULT:
_temp2 temp2=
_temp1 temp1
_temp temp
_ one
----------

----------
LOGQL_EXPRESSION:
this is an url <url> in text
EVENT:
this is an url https://user:password@wazuh.com:8080/path?query=%22a%20query%20with%20a%20space%22#fragment in text
RESULT:
url.query query=%22a%20query%20with%20a%20space%22
url.scheme https
url.username user
url.password password
url.port 8080
url.original https://user:password@wazuh.com:8080/path?query=%22a%20query%20with%20a%20space%22#fragment
url.fragment fragment
url.path /path
url.domain wazuh.com
----------

----------
LOGQL_EXPRESSION:
<source.ip> - <server.ip> -- <source.nat.ip> "-" "-"
EVENT:
127.0.0.1 - 2001:db8:3333:4444:CCCC:DDDD:EEEE:FFFF -- 0:db8:0:4:CCCC:0:EEEE:FFFF "-" "-"
RESULT:
source.nat.ip 0:db8:0:4:CCCC:0:EEEE:FFFF
server.ip 2001:db8:3333:4444:CCCC:DDDD:EEEE:FFFF
source.ip 127.0.0.1
----------

root@ubuntu:/home/palacios/Workspace/wazuh/src/engine/build/source/hlp# ./hlp_cli 2> /dev/null
Usage:
  "FILENAME"
  "LOGQL_EXPRESSION" "EVENT"
root@ubuntu:/home/palacios/Workspace/wazuh/src/engine/build/source/hlp# ./hlp_cli 2> /dev/null
Usage:
 ./hlp_cli"FILENAME"
 ./hlp_cli"LOGQL_EXPRESSION" "EVENT"
root@ubuntu:/home/palacios/Workspace/wazuh/src/engine/build/source/hlp# ./hlp_cli 2> /dev/null
Usage:
 ./hlp_cli "FILENAME"
 ./hlp_cli "LOGQL_EXPRESSION" "EVENT"
root@ubuntu:/home/palacios/Workspace/wazuh/src/engine/build/source/hlp# ./hlp_cli 2> /dev/null "example.yml"
----------
LOGQL_EXPRESSION:
<_map/MAP/ /=/.> <_dummy>
EVENT:
key1=Value1 Key2=Value2. dummy
RESULT:
_dummy dummy
_map {"key1":"Value1","Key2":"Value2"}
----------

----------
LOGQL_EXPRESSION:
<_map/MAP/ /=>
EVENT:
key1=Value1 Key2=
RESULT:
----------

----------
LOGQL_EXPRESSION:
<_json1/JSON> - <_json2/JSON>
EVENT:
{"String":"This is a string"} - {"String":"This is another string"}
RESULT:
_json2 {"String":"This is another string"}
_json1 {"String":"This is a string"}
----------

----------
LOGQL_EXPRESSION:
<_> <_temp> <_temp1/type> <_temp2/type/type2>
EVENT:
one temp temp1 temp2=
RESULT:
_temp2 temp2=
_temp1 temp1
_temp temp
_ one
----------

----------
LOGQL_EXPRESSION:
this is an url <url> in text
EVENT:
this is an url https://user:password@wazuh.com:8080/path?query=%22a%20query%20with%20a%20space%22#fragment in text
RESULT:
url.query query=%22a%20query%20with%20a%20space%22
url.scheme https
url.username user
url.password password
url.port 8080
url.original https://user:password@wazuh.com:8080/path?query=%22a%20query%20with%20a%20space%22#fragment
url.fragment fragment
url.path /path
url.domain wazuh.com
----------

----------
LOGQL_EXPRESSION:
<source.ip> - <server.ip> -- <source.nat.ip> "-" "-"
EVENT:
127.0.0.1 - 2001:db8:3333:4444:CCCC:DDDD:EEEE:FFFF -- 0:db8:0:4:CCCC:0:EEEE:FFFF "-" "-"
RESULT:
source.nat.ip 0:db8:0:4:CCCC:0:EEEE:FFFF
server.ip 2001:db8:3333:4444:CCCC:DDDD:EEEE:FFFF
source.ip 127.0.0.1
----------
```

### CODE REVIEW

**Comment by gdiazlo** [comment]
Please use the testing library Google Test for this kind of testing. All test must be automated and other developers and the CI/CD pipeline might need to run them in a standardized way.

**Comment by palaciosjeremias** [comment]
> Please use the testing library Google Test for this kind of testing. All test must be automated and other developers and the CI/CD pipeline might need to run them in a standardized way.

This tool was implemented for exploratory testing. It allows a tester to pass multiple and diverse formats and events to show the output.
Is complementary to the UT, already implemented.

File: src/engine/source/hlp/tools/hlp_cli.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        auto parseOp = getParserOp(exp_it->c_str());
+        auto result = parseOp(event_it->c_str());
+
+        std::cout << "----------" << std::endl;
+        std::cout << "LOGQL_EXPRESSION:" << std::endl;
+        std::cout << *exp_it << std::endl;
+        std::cout << "EVENT:" << std::endl;
+        std::cout << *event_it << std::endl;
+        std::cout << "RESULT:" << std::endl;
+        for(auto it = result.cbegin(); it != result.cend(); ++it)
```
Comment (san1526) [inline src/engine/source/hlp/tools/hlp_cli.cpp:46]: could use 
```
    fprintf(stderr, "\n%30s | %s\n", "Key", "Val");
    fprintf(stderr, "-------------------------------|------------\n");
    for (auto const &r : result)
    {
        fprintf(stderr, "%30s | %s\n", r.first.c_str(), r.second.c_str());
    }
```
to see the results better

File: src/engine/source/hlp/CMakeLists.txt [hunk 2]
```diff
@@ -48,3 +49,13 @@ target_include_directories(hlp PUBLIC
...
+
+####################################################################################################
+# Test tool
+####################################################################################################
+
+add_executable(hlp_cli
+  ${ENGINE_HLP_TOOLS_DIR}/hlp_cli.cpp
+)
+
+target_link_libraries(hlp_cli hlp ${EXTERNAL_DEPS})
```
Comment (LucioDonda) [inline src/engine/source/hlp/CMakeLists.txt:61]: add empty newline at the end of the file

File: src/engine/source/hlp/tools/hlp_cli.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            std::cout << it->first << " " << it->second << std::endl;
+        }
+        std::cout << "----------" << std::endl << std::endl;
+        exp_it++;
+        event_it++;
+    }
+
+    return 0;
+}
```
Comment (LucioDonda) [inline src/engine/source/hlp/tools/hlp_cli.cpp:56]: add newline at the end.

File: src/engine/source/hlp/tools/hlp_cli.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                logql_expressions.emplace_back(input["logql_expression"].as<std::string>());
+                events.emplace_back(input["event"].as<std::string>());
+            }
+        }
+        catch (const std::exception & e) {
+            std::cout << "Error reading file " << argv[1] << ".Error: " << e.what() << std::endl;
+            return 0;
+        }
+    }
+    else if (argc >= 3) {
```
Comment (LucioDonda) [inline src/engine/source/hlp/tools/hlp_cli.cpp:29]: should we handle errors for more than one pair of event expression (when argc is more than 3) ?

File: src/engine/source/hlp/tools/hlp_cli.cpp (hunk 1 above)
Comment (LucioDonda) [reply src/engine/source/hlp/tools/hlp_cli.cpp:46]: +1

**Review by LucioDonda — APPROVED** [review]
those extra lines should be added but besides that there's only personal comments

File: src/engine/source/hlp/tools/hlp_cli.cpp (hunk 1 above)
Comment (palaciosjeremias) [reply src/engine/source/hlp/tools/hlp_cli.cpp]: Changed!

File: src/engine/source/hlp/CMakeLists.txt (hunk 2 above)
Comment (palaciosjeremias) [reply src/engine/source/hlp/CMakeLists.txt]: Fixed

File: src/engine/source/hlp/tools/hlp_cli.cpp (hunk 3 above)
Comment (palaciosjeremias) [reply src/engine/source/hlp/tools/hlp_cli.cpp]: Fixed

File: src/engine/source/hlp/tools/hlp_cli.cpp (hunk 4 above)
Comment (palaciosjeremias) [reply src/engine/source/hlp/tools/hlp_cli.cpp]: Now is controlled, for clarity

**Pull request merged**